### PR TITLE
feat(ScanCodeResultParser): Warn about unsupported versions

### PR DESCRIPTION
**Pull request by sschuberth**
While parsing newer output format versions might not throw exceptions, results may be incomplete (see #6617 for the context). Avoid confusion by warning in that case.

### CODE REVIEW

File: scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt [hunk 1]
```diff
@@ -101,6 +105,17 @@ internal fun generateSummary(
 ): ScanSummary {
     val header = result["headers"].single()
 
+    val issues = mutableListOf<Issue>()
+    val outputFormatVersion = header["output_format_version"]?.textValue()?.let { Semver(it) }
+    if (outputFormatVersion != null && outputFormatVersion > Semver(MAX_SUPPORTED_OUTPUT_FORMAT_VERSION)) {
+        issues += ScanCode.createAndLogIssue(
+            source = ScanCode.SCANNER_NAME,
+            message = "The output format version $outputFormatVersion exceeds the supported version " +
```
Comment (fviernau) [inline scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt:113]: Should we add ScanCode into the message, e.g. "The **ScanCode** output format version..."?

File: scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt [hunk 2]
```diff
@@ -42,6 +44,8 @@ import org.ossreviewtoolkit.utils.spdx.toSpdxId
 
 import org.semver4j.Semver
 
+const val MAX_SUPPORTED_OUTPUT_FORMAT_VERSION = "2.0.0"
```
Comment (fviernau) [inline scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt:47]: Should this remain internal?

File: scanner/src/test/kotlin/scanners/scancode/ScanCodeResultParserTest.kt [hunk 3]
```diff
@@ -211,6 +216,48 @@ class ScanCodeResultParserTest : FreeSpec({
                 )
             }
         }
+
+        "for output format 3.0.0 should" - {
```
Comment (fviernau) [inline scanner/src/test/kotlin/scanners/scancode/ScanCodeResultParserTest.kt:220]: Consider below alternative which (1) tests both boundary cases, uses (2) `MAX_SUPPORTED_OUTPUT_FORMAT_VERSION` so the test needs not be updated when 
the supported output version is changed and (3) weakens the assertion so that changes to the error message do not require updating the test.

```
"generateSummary()" should {
     "not create a warning for highest supported output version" {
     	val summary = createOutputJson(SpdxConstants.NONE,MAX_SUPPORTED_OUTPUT_FORMAT_VERSION))
        summary.issues.count { it.severity == Severity.WARNING } shouldBe 0
     }
     
     "create a warning for lowest unsupported output version" {
        val version = Semver(MAX_SUPPORTED_OUTPUT_FORMAT_VERSION).withIncPatch()
     	val summary = createOutputJson(SpdxConstants.NONE, version.toString())
     
        summary.issues.count { it.severity == Severity.WARNING } shouldBe 1
     }
}

fun createOutputJson(outputFormatVersion: String): JsonNode {
    ...
}
```

File: scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt (hunk 1 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt:113]: I believe the `source` is sufficient here.

File: scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt (hunk 2 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt:47]: I was considering that, but then though programmatic users of the scanner library might want to access this.

File: scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt (hunk 1 above)
Comment (fviernau) [reply scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt:113]: The source cannot be matched  by resolutions...So, just from the message the scanner name cannot be matched. Anyhow, we also do not have this error for other scanners right now.

File: scanner/src/test/kotlin/scanners/scancode/ScanCodeResultParserTest.kt (hunk 3 above)
Comment (sschuberth) [reply scanner/src/test/kotlin/scanners/scancode/ScanCodeResultParserTest.kt:220]: > "not create a warning for highest supported output version"

I believe that case is sufficiently covered by other tests.

> "create a warning for lowest unsupported output version"

This actually pointed me at the fact that we should consider less-than-major changes in the output format version compatible. I'll change the implementation accordingly.

File: scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt (hunk 1 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/scancode/ScanCodeResultParser.kt]: BTW, this generally relates to https://github.com/oss-review-toolkit/ort/issues/4393.

**Review by fviernau — APPROVED** [review]

**Comment by sschuberth** [comment]
Merging despite the unrelated `BabelFunTest` failure.
